### PR TITLE
fixed bug in trie.extensionNode when splitting into a branch

### PR DESCRIPTION
**Pull request by iulianpascalau**
added edge case tests and extensive tests that prove consistency between iterations
added pretty print functionality to a trie

### CODE REVIEW

File: data/trie/extensionNode.go [hunk 1]
```diff
@@ -299,9 +300,11 @@ func (en *extensionNode) insert(n *leafNode, db data.DBWriteCacher, marshalizer
 	if childPosOutOfRange(oldChildPos) || childPosOutOfRange(newChildPos) {
 		return false, nil, ErrChildPosOutOfRange
 	}
-	branch.children[oldChildPos] = newExtensionNode(en.Key[keyMatchLen+1:], en.child)
-	if len(en.Key) == 1 {
+	followingExtensionNode := newExtensionNode(en.Key[keyMatchLen+1:], en.child)
```
Comment (sasurobert) [inline data/trie/extensionNode.go]: good catch +1

File: data/trie/patriciaMerkleTrie.go [hunk 2]
```diff
@@ -241,6 +243,20 @@ func (tr *patriciaMerkleTrie) Recreate(root []byte) (data.Trie, error) {
...
 }
 
+// Print outputs a graphical view of the trie. Mainly used in tests/debugging
+func (tr *patriciaMerkleTrie) Print(writer io.Writer) {
+	if writer == nil {
+		return
+	}
+	if tr.root == nil {
+		_, _ = fmt.Fprintln(writer, "*** EMPTY TRIE ***")
+
```
Comment (sasurobert) [inline data/trie/patriciaMerkleTrie.go]: delete empty line

File: data/trie/extensionNode.go (hunk 1 above)
Comment (sasurobert) [inline data/trie/extensionNode.go]: add an empty line before

File: integrationTests/state/genesisBad.json [hunk 3]
```diff
@@ -0,0 +1,28 @@
+{
```
Comment (sasurobert) [inline integrationTests/state/genesisBad.json]: rename file: edgeCase ?

File: integrationTests/state/genesisState_test.go [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+	fmt.Printf("root hash2: %s\n", base64.StdEncoding.EncodeToString(hash2))
+
+	fmt.Println()
+	tr2.Print(os.Stdout)
+	fmt.Println()
+
+	assert.Equal(t, hash1, hash2)
+}
+
+func TestExtensionNodeToBranchVariant2(t *testing.T) {
```
Comment (sasurobert) [inline integrationTests/state/genesisState_test.go]: what does variant1 and variant2 mean ? :)

File: integrationTests/state/genesisState_test.go [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+	tr2.Print(os.Stdout)
+	fmt.Println()
+
+	hash2, _ := tr2.Root()
+	fmt.Printf("root hash2: %s\n", base64.StdEncoding.EncodeToString(hash2))
+
+	assert.Equal(t, hash1, hash2)
+}
+
+func TestExtensiveUpdatesAndRemovesWithConsistencyBetweenCylces(t *testing.T) {
```
Comment (sasurobert) [inline integrationTests/state/genesisState_test.go]: would move to a smokeTest folder. suggestion.

File: data/trie/patriciaMerkleTrie.go (hunk 2 above)
Comment (iulianpascalau) [reply data/trie/patriciaMerkleTrie.go]: Refactored whole method

File: data/trie/extensionNode.go (hunk 1 above)
Comment (iulianpascalau) [reply data/trie/extensionNode.go]: done

File: integrationTests/state/genesisBad.json (hunk 3 above)
Comment (iulianpascalau) [reply integrationTests/state/genesisBad.json]: done genesisEdgeCase.json

File: integrationTests/state/genesisState_test.go (hunk 4 above)
Comment (iulianpascalau) [reply integrationTests/state/genesisState_test.go]: changed from variantX to EdgeCaseX as I have found out 2 edge cases

File: integrationTests/state/genesisState_test.go (hunk 5 above)
Comment (iulianpascalau) [reply integrationTests/state/genesisState_test.go]: TBD. Will move other tests (in another PR) as well if we decide to do that.

File: data/trie/patriciaMerkleTrie.go [hunk 6]
```diff
@@ -241,6 +242,19 @@ func (tr *patriciaMerkleTrie) Recreate(root []byte) (data.Trie, error) {
 	return newTr, nil
 }
 
+// String outputs a graphical view of the trie. Mainly used in tests/debugging
```
Comment (raduchis) [inline data/trie/patriciaMerkleTrie.go]: Because it is mainly used in tests/debugging, wouldn't it be better to have a different printer of a tree somewhere? This way, code that is not usefull is encapsulated in a different component. This printer will also know the printfunctions for the LeafNode, BranchNode...
